url5997ccb760_JANNLab_JANNLab_tgz-pJ8-de_jannlab_examples_recurrent_SequentialXorExampleJ8/src/de/jannlab/training/GradientDescent.java

```diff
@@ -0,0 +1,489 @@
+     * values.
+     */
+    private void computeWeightDiffs() {
+        //
+        // add momentum and multiply learning rate mu. 
+        //
+        for (int i = 1; i <= this.weightsnum; i++) {
+            //
+            // calculate momentum.
+            //
+            final double mw = this.dweightslast[i] * this.alpha; 
+            this.dweights[i] = (this.mu * this.dweights[i]) + mw;
+        }  
+    }
+    
+    /**
+     * Adjusts the weights.
+     */
+    private void adjustWeights() {
+        //
+        for (int i = 1; i <= this.weightsnum; i++) {
+            //
+            this.weights[i] += this.dweights[i];
+            /*
+            // # DEBUG #
+            this.weightsdbg[i] += Math.abs(this.dweights[i]);
+            // #
+            */
+        }
+        
+    }
+    
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    synchronized
+    public void train() {
+        //
+        this.init();
+        //
+        this.notifyStarted();
+        //
+        // iterate epochs.
+        //
+        final SampleSet tset = this.trainset;
+        final int tsetsize   = this.permutation.length;
+        //
+        int count           = 0;
+        int nbetterctr      = 0;
+        double[] minweights = this.net.getWeights().clone();
+        double minerror = Double.MAX_VALUE;
+        //
+        for (int i = 0; i < this.epochs; i++) {
+            this.epoch = i;
+            //
+            double epocherror = 0.0;
+            if (this.permute) {
+                IntTools.shuffle(this.permutation, this.rnd);
+            }
+            //
+            // do online or offline gradient descent?
+            //
+            if (this.online) {
+                //
+                // for all patterns in trainset.
+                //
+                for (int j = 0; j < tsetsize; j++) {
+                    //
+                    // determine permuted index and sample.
+                    //
+                    final int idx = this.permutation[j];
+                    final Sample sample = tset.get(idx);
+                    //
+                    // compute forward pass.
+                    //
+                    this.net.reset();
+                    final double err = NetTools.performForward(
+                        this.net, sample, this.features
+                    );
+                    final int frameidx = this.net.getFrameIdx();
+                    //
+                    // compute backward pass.
+                    //
+                    NetTools.performBackward(this.net);
+                    //
+                    // compute weight differences and adjust weights.
+                    //
+                    this.resetWeightDiffs();
+                    this.accumulateWeightsDiffs(frameidx);
+                    this.computeWeightDiffs();
+                    this.adjustWeights();
+                    //
+                    epocherror += err;
+                }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 384:411, 412:444.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.